syntax = "proto3";
option go_package = "../main";
package calculator;


message ArithmeticOpArguments {
  int32 arg1 = 1;
  int32 arg2 = 2;
}

message SingleIntResult {
  int32 res = 1;
}

service Calculator {
  rpc SimpleAdd (ArithmeticOpArguments) returns (SingleIntResult) {}
  rpc SimpleSubtract (ArithmeticOpArguments) returns (SingleIntResult) {}
}

enum OperationType {
  SUM = 0;
  AVG = 1;
  MIN = 2;
  MAX = 3;
}

message ComplexArithmeticOpArguments {
  OperationType optype = 1;  
  repeated double args = 2;
}

message ComplexArithmeticOpResult {
  double res = 1;
}

message ListArithmeticOpArguments {
  repeated int32 args = 2;
}
     
service AdvancedCalculator {
  rpc ComplexOperation (ComplexArithmeticOpArguments) returns (ComplexArithmeticOpResult) {}
  rpc ListSum (ListArithmeticOpArguments) returns (SingleIntResult) {}
}